django/contrib/admin/sites.py

```diff
@@ -409,6 +420,35 @@ def login(self, request, extra_context=None):
         request.current_app = self.name
         return LoginView.as_view(**defaults)(request)
 
+    def final_catch_all_view(self, request):
+        # TODO: Remove the if block and the _admin_final_catch_all_view
+        # attribute after the deprecation period ends.
+        if (
+            request.user.is_authenticated and
+            settings.APPEND_SLASH and
+            not request.path_info.endswith('/')
+        ):
+            urlconf = getattr(request, 'urlconf', None)
+            path = '%s/' % request.path_info
+            try:
+                match = resolve(path, urlconf)
+            except Resolver404:
+                pass
+            else:
+                if not getattr(match.func, '_admin_final_catch_all_view', False):
+                    warnings.warn(
+                        "The path %s was requested with setting APPEND_SLASH "
+                        "set to True, so the response will be a redirect to %s. "
+                        "Support for APPEND_SLASH with the admin is deprecated "
+                        "and will be removed in Django 4.1 to prevent a private "
+                        "data leak issue. An Http404 will be raised instead." % (request.path_info, path),
+                        RemovedInDjango41Warning,
+                    )
```

Comment:
Can we please exclude any potential deprecation of the APPEND_SLASH in the admin from the scope of this PR.

Reply:
Doing so would mean we would continue to support behavior that we know can leak private URLs and private backend details. These are private details that some people -- including myself -- have concern about in their applications. That is something that I would prefer not to have my name on. Django has a history of being secure and private by default, I think that applies here as well. If this current code change is not acceptable, perhaps we can discuss one that is that still moves towards private by default.

If you have a concrete suggestion that would both stop the leaking of private information as well as satisfy the backward compatibility concerns, please let me know.